### PR TITLE
PERF-#6378: use numpy.array instead of list in internals of iloc/loc operation

### LINKED ISSUE #6378: use numpy.array instead of list in internals of iloc/loc operation

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

numpy arrays are much faster to serialize than lists , what will need to be done next when performing `map_axis_partitions` operation.

<details>

<summary>
Script to measure
</summary>

```python
import numpy as np
import ray
from time import time

# initialize it
ray.put([1,2,3,4])

np.random.seed(42)

len_rows = 10**7
sorted_rows_position = np.arange(len_rows)
row_positions = sorted_rows_position.copy()
np.random.shuffle(row_positions)


row_order_mapping = dict(
    zip(sorted_rows_position, range(len(row_positions)))
)

print("old way")
for _ in range(3):
    start = time()
    new_row_order = [row_order_mapping[idx] for idx in row_positions]
    print(f"time: {time()-start}")


start = time()
ref = ray.put(new_row_order)
print(f"ray put {time()-start}")
del ref


print("from_iter way")
for _ in range(3):
    start = time()
    new_row_order = np.fromiter((row_order_mapping[idx] for idx in row_positions), dtype="int64", count=len(row_positions))
    print(f"time: {time()-start}")

start = time()
ref = ray.put(new_row_order)
print(f"ray put {time()-start}")
```

</details>

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6378 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by anmyachev** [comment]
@dchigarev please review

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -1110,14 +1110,22 @@ def _maybe_reorder_labels(
             row_order_mapping = dict(
                 zip(sorted_row_positions, range(len(row_positions)))
             )
-            new_row_order = [row_order_mapping[idx] for idx in row_positions]
+            new_row_order = np.fromiter(
+                (row_order_mapping[idx] for idx in row_positions),
+                dtype="int64",
+                count=len(row_positions),
+            )
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1117]: from the surrounding code, I believe we can make the following:
1. `sorted_row_positions` is a numpy array (we can make it 100% sure by calling `numpy.asarray()` which is a no-op for a thing which is already an array)
2. same for `row_positions`
3. then making the dictionary and then building an array via `np.fromiter` is just a super-slow version of `sorted_row_positions[row_positions]` 😉 

If I'm right, this piece could be sped up even further (and code greatly simplified) by rewriting as
```python
new_row_order = np.asarray(sorted_row_positions)[np.asarray(row_positions)]
```

Comment (vnlitvinov) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1117]: What does it matter if it's a contiguous array (if you mean in-memory contiguous)? Numpy is smart enough:
```python
>>> np.array([1,5,7,2,9,12,19,-5,6])[np.array([2,6])]
array([ 7, 19])
>>> x1=np.array(np.arange(10))
>>> x1.flags
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

>>> x2=x1[::-1]
>>> x2.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

>>> x2[np.array([2,3])]
array([7, 6])
>>> x1[x2]
array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])
```

Comment (anmyachev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1117]: I have the following example in my mind:
```python
>>> import numpy as np
>>> row_positions = np.array([7,1,3])
>>> sorted_row_positions = np.array([1,3,7])
>>> sorted_row_positions[row_positions]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index 7 is out of bounds for axis 0 with size 3
>>>
```

Comment (vnlitvinov) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1117]: I have built an alternative approach, need to benchmark it before making a PR though...

**Pull request closed without merging**